### PR TITLE
Step 4 SSR

### PR DESCRIPTION
**Pull request by a-reznikov**
- [x] Read about SSR
- [x] Read how NextJS work with that -[ https://nextjs.org/docs/basic-features/data-fetching/overview](https://nextjs.org/docs/basic-features/data-fetching/overview) and other pages on[ nextjs.org/docs](http://nextjs.org/docs) about it
- [x] Implement SSR on landing page 
- [x] Check how it working and what the difference between CSR and SSR
- [x] Create commit and PR to main branch

### CODE REVIEW

File: src/components/news-list/NewsList.tsx [hunk 1]
```diff
@@ -1,39 +1,35 @@
...
-            "md:hover:scale-105"
-          )}
-        >
-          <Link href={`${section}/details/${news.title}`}>
-            <NewsCard news={news} index={index} />
-          </Link>
-        </li>
-      ))}
+      {data &&
+        data.map((news, index) => (
```
Comment (Ilya-Valasiuk) [inline src/components/news-list/NewsList.tsx]: data?.map

File: src/pages/index.tsx [hunk 2]
```diff
@@ -1,4 +1,17 @@
+import { dehydrate } from "@tanstack/react-query";
 import { NewsList } from "@/components/news-list/NewsList";
+import { prefetchNews } from "@/api/news/queries";
+import { QUERY_CLIENT } from "@/constants/query-client";
+
+export async function getStaticProps() {
```
Comment (Ilya-Valasiuk) [inline src/pages/index.tsx]: it's error

File: src/pages/[section]/index.tsx [hunk 3]
```diff
@@ -1,15 +1,33 @@
...
+export const getServerSideProps: GetServerSideProps<Props> = async (
+  context
+) => {
+  const section = hasSlugStringType(context.query.section)
+    ? context.query.section
     : "";
 
-  return <NewsList section={sectionName} />;
+  await prefetchNews(section);
+
```
Comment (Ilya-Valasiuk) [inline src/pages/[section]/index.tsx:23]: if (!sectrion) -> not found

File: src/api/articles/queries.ts [hunk 4]
```diff
@@ -1,14 +1,21 @@
 import { useQuery } from "@tanstack/react-query";
+import { QUERY_CLIENT } from "@/constants/query-client";
 import { searchArticles } from "./module";
 
 enum QueryKeys {
```
Comment (Ilya-Valasiuk) [inline src/api/articles/queries.ts:5]: ?

File: src/api/news/queries.ts [hunk 5]
```diff
@@ -1,14 +1,21 @@
 import { useQuery } from "@tanstack/react-query";
+import { QUERY_CLIENT } from "@/constants/query-client";
 import { getNewsBySection } from "./module";
 
-enum QueryKeys {
+export enum QueryKeys {
```
Comment (Ilya-Valasiuk) [inline src/api/news/queries.ts]: ?

File: src/api/news/queries.ts [hunk 6]
```diff
@@ -1,14 +1,21 @@
...
 export const useNews = (sectionName: string) => {
   return useQuery({
     queryKey: [QueryKeys.SECTION, sectionName],
     queryFn: () => getNewsBySection(sectionName),
-    enabled: !!sectionName,
+  });
+};
+
+export const prefetchNews = async (sectionName: string) => {
+  return QUERY_CLIENT.prefetchQuery({
```
Comment (Ilya-Valasiuk) [inline src/api/news/queries.ts]: QUERY_CLIENT -> queryClient

File: src/api/news/queries.ts [hunk 7]
```diff
@@ -1,14 +1,21 @@
...
 
 export const useNews = (sectionName: string) => {
   return useQuery({
     queryKey: [QueryKeys.SECTION, sectionName],
     queryFn: () => getNewsBySection(sectionName),
-    enabled: !!sectionName,
+  });
+};
+
+export const prefetchNews = async (sectionName: string) => {
```
Comment (Ilya-Valasiuk) [inline src/api/news/queries.ts]:
```
export const prefetchNews = async (sectionName: string) => 
  queryProvider.prefetchQuery({
    queryKey: [QueryKeys.SECTION, sectionName],
    queryFn: () => getNewsBySection(sectionName),
  });
```

File: src/components/news-list/NewsList.tsx (hunk 1 above)
Comment (a-reznikov) [reply src/components/news-list/NewsList.tsx]: Used

File: src/pages/index.tsx (hunk 2 above)
Comment (a-reznikov) [reply src/pages/index.tsx]: Used getServerSideProps

File: src/pages/[section]/index.tsx (hunk 3 above)
Comment (a-reznikov) [reply src/pages/[section]/index.tsx:23]: Implemented

File: src/api/news/queries.ts (hunk 5 above)
Comment (a-reznikov) [reply src/api/news/queries.ts]: Deleted

File: src/api/news/queries.ts (hunk 6 above)
Comment (a-reznikov) [reply src/api/news/queries.ts]: Changed

File: src/api/news/queries.ts (hunk 7 above)
Comment (a-reznikov) [reply src/api/news/queries.ts]: Applied

File: src/pages/[section]/details/[title].tsx [hunk 8]
```diff
@@ -1,12 +1,38 @@
...
+
+export const getServerSideProps: GetServerSideProps<Props> = async (
+  context
+) => {
+  const title = hasSlugStringType(context.query.title)
+    ? context.query.title
+    : "";
+
+  try {
+    await prefetchArticles(title);
```
Comment (Ilya-Valasiuk) [inline src/pages/[section]/details/[title].tsx:22]: if title doesn't exist - it doesn't make sense to make call, right?

So why not to add check like

```
if (!title) {
  throw new Error('title missed')
}
```

File: src/pages/[section]/index.tsx (hunk 3 above)
Comment (Ilya-Valasiuk) [reply src/pages/[section]/index.tsx:23]: not really - you make API call with empty parameter. why?

File: src/pages/_app.tsx [hunk 9]
```diff
@@ -1,18 +1,27 @@
-import { QueryClientProvider } from "@tanstack/react-query";
+import { QueryClientProvider, HydrationBoundary } from "@tanstack/react-query";
 import type { AppProps } from "next/app";
-import { QUERY_CLIENT } from "@/constants/query-client";
+import { useRouter } from "next/router";
+import { queryClient } from "@/constants/query-client";
 import { RootLayout } from "@/layouts/root-layout/RootLayout";
 import { OPEN_SANS } from "@/fonts";
 import "@/styles/globals.css";
+import { hasSlugStringType } from "@/utils/type-guards";
```
Comment (Ilya-Valasiuk) [inline src/pages/_app.tsx:7]: styles always last in import order

File: src/pages/[section]/index.tsx (hunk 3 above)
Comment (a-reznikov) [reply src/pages/[section]/index.tsx:23]: Implemented throwing an error when missing the slug and returning notFound for other errors in the response from the API.

File: src/pages/[section]/details/[title].tsx (hunk 8 above)
Comment (a-reznikov) [reply src/pages/[section]/details/[title].tsx:22]: Implemented throwing an error when missing the slug and returning notFound for other errors in the response from the API.

File: src/pages/_app.tsx (hunk 9 above)
Comment (a-reznikov) [reply src/pages/_app.tsx:7]: Fixed

**Pull request merged**